File: rename_chapters.mjs
#!/usr/bin/env node
// -*- coding: utf-8 -*-
/**
 * 从章纲文件中提取章节标题，批量重命名正文目录下的章节文件。
 * 重命名规则：第X章-标题.md
 */

import fs from 'fs';
import path from 'path';

// 目录配置
const ZHANGGANG_DIR = "3-大纲/章纲";
const ZHENGWEN_DIR = "4-正文/第一卷  初来乍到";

/**
 * 从所有章纲文件中提取 {章节号: 标题} 映射
 */
function extractChapterTitles(zhanggangDir) {
    const chapterMap = {};

    // 中文数字映射
    const cnNumMap = {
        '一': 1, '二': 2, '三': 3, '四': 4, '五': 5,
        '六': 6, '七': 7, '八': 8, '九': 9, '十': 10,
        '十一': 11, '十二': 12, '十三': 13, '十四': 14, '十五': 15,
        '十六': 16, '十七': 17, '十八': 18, '十九': 19, '二十': 20,
        '二十一': 21, '二十二': 22, '二十三': 23, '二十四': 24, '二十五': 25,
        '二十六': 26, '二十七': 27, '二十八': 28, '二十九': 29, '三十': 30,
        '三十一': 31, '三十二': 32, '三十三': 33, '三十四': 34, '三十五': 35,
        '三十六': 36, '三十七': 37, '三十八': 38, '三十九': 39, '四十': 40,
    };

    // 匹配 "## 第X章：标题" 格式（支持阿拉伯数字和中文数字）
    const patternArabic = /^##\s+第(\d+)章[：:]\s*(.+?)$/;
    const patternChinese = /^##\s+第([一二三四五六七八九十]+)章[：:]\s*(.+?)$/;

    // 获取章纲目录下所有 .md 文件
    const files = fs.readdirSync(zhanggangDir)
        .filter(f => f.endsWith('.md'))
        .sort();

    for (const file of files) {
        const filepath = path.join(zhanggangDir, file);
        const content = fs.readFileSync(filepath, 'utf-8');
        const lines = content.split('\n');

        for (const line of lines) {
            const trimmed = line.trim();
            // 尝试匹配阿拉伯数字
            let m = trimmed.match(patternArabic);
            if (m) {
                const chapterNum = parseInt(m[1], 10);
                const title = m[2].trim();
                chapterMap[chapterNum] = title;
                continue;
            }
            // 尝试匹配中文数字
            m = trimmed.match(patternChinese);
            if (m) {
                const cnStr = m[1];
                if (cnNumMap[cnStr] !== undefined) {
                    const chapterNum = cnNumMap[cnStr];
                    const title = m[2].trim();
                    chapterMap[chapterNum] = title;
                }
            }
        }
    }

    return chapterMap;
}

/**
 * 根据 {章节号: 标题} 映射，重命名正文目录下的文件
 */
function renameFiles(zhengwenDir, chapterMap) {
    // 匹配 "第X章.md" 格式
    const filePattern = /^第(\d+)章\.md$/;

    const renamed = [];
    const errors = [];

    const dirContents = fs.readdirSync(zhengwenDir);

    for (const filename of dirContents) {
        const m = filename.match(filePattern);
        if (!m) continue;

        const chapterNum = parseInt(m[1], 10);

        if (!(chapterNum in chapterMap)) {
            errors.push(`第${chapterNum}章：章纲中未找到对应标题`);
            continue;
        }

        const title = chapterMap[chapterNum];

        // 清理标题中的非法字符（Windows文件名不允许的字符）
        const safeTitle = title.replace(/[<>:"/\\|?*]/g, '');
        const newName = `第${chapterNum}章-${safeTitle}.md`;

        const oldPath = path.join(zhengwenDir, filename);
        const newPath = path.join(zhengwenDir, newName);

        if (fs.existsSync(newPath)) {
            errors.push(`第${chapterNum}章：目标文件已存在 ${newName}`);
            continue;
        }

        fs.renameSync(oldPath, newPath);
        renamed.push(`第${chapterNum}章 → 第${chapterNum}章-${title}.md`);
    }

    return { renamed, errors };
}

function main() {
    console.log("=".repeat(60));
    console.log("从章纲提取章节标题...");
    const chapterMap = extractChapterTitles(ZHANGGANG_DIR);

    const sortedChapters = Object.entries(chapterMap).sort((a, b) => a[0] - b[0]);

    if (sortedChapters.length === 0) {
        console.log("错误：未找到任何章节标题！");
        return;
    }

    console.log(`成功提取 ${sortedChapters.length} 个章节标题（展示前10个和后3个）：`);
    for (const [num, title] of sortedChapters.slice(0, 10)) {
        console.log(`  第${num}章: ${title}`);
    }
    if (sortedChapters.length > 10) {
        console.log(`  ... 共${sortedChapters.length}个 ...`);
        for (const [num, title] of sortedChapters.slice(-3)) {
            console.log(`  第${num}章: ${title}`);
        }
    }

    console.log("\n" + "=".repeat(60));
    console.log("处理正文目录文件...");

    const { renamed, errors } = renameFiles(ZHENGWEN_DIR, chapterMap);

    console.log(`\n成功重命名 ${renamed.length} 个文件：`);
    for (const r of renamed) {
        console.log(`  ✓ ${r}`);
    }

    if (errors.length > 0) {
        console.log(`\n错误 ${errors.length} 个：`);
        for (const e of errors) {
            console.log(`  ✗ ${e}`);
        }
    }

    console.log("\n" + "=".repeat(60));
    console.log("完成！");
}

main();

File: rename_chapters.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
从章纲文件中提取章节标题，批量重命名正文目录下的章节文件。
重命名规则：第X章-标题.md
"""

import os
import re
import glob

# 目录配置
ZHANGGANG_DIR = "3-大纲/章纲"         # 章纲目录
ZHENGWEN_DIR = "4-正文/第一卷  初来乍到"  # 正文目录

def extract_chapter_titles(zhanggang_dir):
    """从所有章纲文件中提取 {章节号: 标题} 映射"""
    chapter_map = {}
    
    # 匹配 "## 第X章：标题" 或 "## 第X章：标题" 格式
    pattern = re.compile(r'^##\s+第(\d+)章[：:]\s*(.+?)$')
    
    # 遍历所有章纲文件
    zhanggang_files = sorted(glob.glob(os.path.join(zhanggang_dir, "*.md")))
    
    for filepath in zhanggang_files:
        with open(filepath, 'r', encoding='utf-8') as f:
            for line in f:
                m = pattern.match(line.strip())
                if m:
                    chapter_num = int(m.group(1))
                    title = m.group(2).strip()
                    chapter_map[chapter_num] = title
    
    return chapter_map

def rename_files(zhengwen_dir, chapter_map):
    """根据 {章节号: 标题} 映射，重命名正文目录下的文件"""
    
    # 匹配 "第X章.md" 格式
    pattern = re.compile(r'^第(\d+)章\.md$')
    
    renamed = []
    errors = []
    
    for filename in os.listdir(zhengwen_dir):
        m = pattern.match(filename)
        if not m:
            continue
        
        chapter_num = int(m.group(1))
        if chapter_num not in chapter_map:
            errors.append(f"第{chapter_num}章：章纲中未找到对应标题")
            continue
        
        title = chapter_map[chapter_num]
        
        # 新文件名：第X章-标题.md
        # 清理标题中的非法字符
        safe_title = re.sub(r'[<>:"/\\|?*]', '', title)
        new_name = f"第{chapter_num}章-{safe_title}.md"
        
        old_path = os.path.join(zhengwen_dir, filename)
        new_path = os.path.join(zhengwen_dir, new_name)
        
        if os.path.exists(new_path):
            errors.append(f"第{chapter_num}章：目标文件已存在 {new_name}")
            continue
        
        os.rename(old_path, new_path)
        renamed.append(f"第{chapter_num}章 → 第{chapter_num}章-{title}.md")
    
    return renamed, errors

def main():
    print("=" * 60)
    print("从章纲提取章节标题...")
    chapter_map = extract_chapter_titles(ZHANGGANG_DIR)
    
    if not chapter_map:
        print("错误：未找到任何章节标题！")
        return
    
    print(f"成功提取 {len(chapter_map)} 个章节标题：")
    
    # 按章节号排序并分组打印
    sorted_chapters = sorted(chapter_map.items())
    for num, title in sorted_chapters[:10]:
        print(f"  第{num}章: {title}")
    
    if len(sorted_chapters) > 10:
        print(f"  ... (共{len(sorted_chapters)}个，省略了{len(sorted_chapters)-10}个)")
        for num, title in sorted_chapters[-3:]:
            print(f"  第{num}章: {title}")
    
    print("\n" + "=" * 60)
    print("处理正文目录文件...")
    
    renamed, errors = rename_files(ZHENGWEN_DIR, chapter_map)
    
    print(f"\n成功重命名 {len(renamed)} 个文件：")
    for r in renamed:
        print(f"  ✓ {r}")
    
    if errors:
        print(f"\n错误 {len(errors)} 个：")
        for e in errors:
            print(f"  ✗ {e}")
    
    print("\n" + "=" * 60)
    print("完成！")

if __name__ == "__main__":
    main()